### DISCUSSION TITLE
Don’t know why server doesn’t accept aiohttp requests

### DISCUSSION THREAD

**Discussion by Pyprohly** [aio-libs/aiohttp#6023] [Q&A]
The following aiohttp code results in a 429 HTTP error[.][1]

[1]: https://www.reddit.com/r/redditdev/comments/pwyrwb/http_429_when_using_python_aiohttp_but_wget_works/

```py
import asyncio
import aiohttp

url = "http://www.reddit.com/.json?limit=1"
headers = {'User-Agent': 'asdf', 'Connection': 'close'}

async def main():
    async with aiohttp.ClientSession() as session:
        async with session.get(url, headers=headers) as resp:
            print(await resp.text())
            print(resp.status)

asyncio.run(main())
```

But using requests and httpx works fine.

```py
url = "http://www.reddit.com/.json?limit=1"
headers = {'User-Agent': 'asdf', 'Connection': 'close'}

import requests, httpx
for pkg in (requests, httpx):
    resp = pkg.get(url, headers=headers)
    print(resp.text)
    print(resp.status_code)
```

Using command line tools like `curl`, `wget`, or `ncat` also work.

```
curl -L -H 'User-Agent: asdf' 'http://www.reddit.com/.json?limit=1'


wget -qO - --header='User-Agent: asdf' 'https://www.reddit.com/.json?limit=1'


ncat --ssl reddit.com 443 <<EOF
GET /.json?limit=1 HTTP/1.1
Host: www.reddit.com
User-Agent: asdf
Connection: close
Accept: */*
Accept-Encoding: identity

EOF
```

So what is different about aiohttp's request that makes reddit servers not like it and how do I get it to work?

**Comment by Dreamsorcerer** [comment]
Weird. The response says to contact them (if you scroll past all the badgers...), so maybe they can explain why they are rate limiting a single request. The response claims this is done by IP, but that clearly is not true.

**Reply by tailoric** [reply]
I have found that aiohttp on python version 3.6 actually works
using the following script
```py
import aiohttp
import asyncio

headers = {'User-Agent': 'Some custom test header'}
url = "https://www.reddit.com/r/all.json"
newline = "\n"
print (aiohttp.__version__)
async def on_request_start(session, trace_config_ctx, params):
    print(params)
    print(f"""
--------START---------------
{params.method} {params.url}
{newline.join(f'{k}: {v}' for k,v in params.headers.items())}
    """)

trace_config = aiohttp.TraceConfig()
trace_config.on_request_start.append(on_request_start)
async def aiohttp_call():
    async with aiohttp.ClientSession(trace_configs=[trace_config]) as session, session.get(url=url, headers=headers) as resp:
        print(resp.headers)
        resp.raise_for_status()
        print(resp.status)
        print(await resp.json())
try:
    loop = asyncio.get_event_loop()

    loop.run_until_complete(aiohttp_call())
except Exception as e:
    print(e)
```

on python 3.6 works   
> [output3.6.txt](https://github.com/aio-libs/aiohttp/files/7287482/output3.6.txt)
```
3.7.4.post0
TraceRequestStartParams(method='GET', url=URL('https://www.reddit.com/r/all.json'), headers=<CIMultiDict('User-Agent': 'Some custom test header')>)

--------START---------------
GET https://www.reddit.com/r/all.json
User-Agent: Some custom test header
    
<CIMultiDictProxy('Connection': 'keep-alive', 'Content-Length': '37469', 'Content-Type': 'application/json; charset=UTF-8', 'x-ua-compatible': 'IE=edge', 'x-frame-options': 'SAMEORIGIN', 'x-content-type-options': 'nosniff', 'x-xss-protection': '1; mode=block', 'x-ratelimit-remaining': '295', 'x-ratelimit-used': '5', 'x-ratelimit-reset': '117', 'Access-Control-Allow-Origin': '*', 'Access-Control-Expose-Headers': 'X-Moose', 'Content-Encoding': 'gzip', 'Cache-Control': 'max-age=0, must-revalidate', 'X-Moose': 'majestic', 'Accept-Ranges': 'bytes', 'Date': 'Tue, 05 Oct 2021 16:18:04 GMT', 'Via': '1.1 varnish', 'Vary': 'accept-encoding', 'Set-Cookie': 'loid=0000000000f5pz8lhr.2.1633450683500.Z0FBQUFBQmhYSHE3SUdCWkdzVkZKWVh2Q3dOR2RNYl82clRvN0FYdEVuWl9NNjZhVnUzTW5pZ2RGOHFyMm5ZNkQzajNrcV9rUDI5b3dVdjdZM0JHbkp2YTgzcnV5MmotOGpsS0tDcEFKLUtoQ3FMX1d4eEotbWtCc0xpMVN1QVhXMHo5OFcxb0hmeXg; Domain=reddit.com; Max-Age=63071999; Path=/; expires=Thu, 05-Oct-2023 16:18:03 GMT; secure; SameSite=None; Secure', 'Set-Cookie': 'session_tracker=gVr1IxY5BdzA5SeNuB.0.1633450683500.Z0FBQUFBQmhYSHE3Z3dNc2JGMVhmTTFzZ3llN2FoQlk3Tk9Pb0pOMEVVeEtrNHlMNWg1M0lMTUxEejdyNkNkNG5PemFqZDY3b0t5REZPY3hLemRxZjZweDV1cnVYOEw5MFF2VUdpLTdsbm5CVDg4Y2NiRjYtN1loY3lNd2VZUkFKcDVmMFBFSlIzbVQ; Domain=reddit.com; Max-Age=7199; Path=/; expires=Tue, 05-Oct-2021 18:18:03 GMT; secure; SameSite=None; Secure', 'Set-Cookie': 'csv=1; Max-Age=63072000; Domain=.reddit.com; Path=/; Secure; SameSite=None', 'Set-Cookie': 'edgebucket=6cZfHvISPRZpL3SRcr; Domain=reddit.com; Max-Age=63071999; Path=/;  secure', 'Strict-Transport-Security': 'max-age=15552000; includeSubDomains; preload', 'Server': 'snooserv', 'X-Clacks-Overhead': 'GNU Terry Pratchett')>
200
```
on python 3.9:


> [output3.9.txt](https://github.com/aio-libs/aiohttp/files/7287489/output3.9.txt)

```
3.7.4.post0
TraceRequestStartParams(method='GET', url=URL('https://www.reddit.com/r/all.json'), headers=<CIMultiDict('User-Agent': 'Some custom test header')>)

--------START---------------
GET https://www.reddit.com/r/all.json
User-Agent: Some custom test header
    
429, message='Too Many Requests', url=URL('https://www.reddit.com/r/all.json')
```

I have tested with `3.6`, `3.7`, `3.8` and `3.9` 
only `3.6` gave a 200 back

**Reply by Dreamsorcerer** [reply]
Can you check exactly what is being sent on the HTTP connections for 3.6 and 3.9? See if there are any differences in the HTTP request.

e.g. You can run `nc -l 8000` (e.g. installed in Ubuntu with the `netcat-openbsd` package) and then change the URL to `http://localhost:8000` and run the example.

**Reply by tailoric** [reply]
with python 3.6:
```
GET / HTTP/1.1
Host: localhost:8000
User-Agent: Some custom test header
Accept: */*
Accept-Encoding: gzip, deflate
```

with python 3.9
```
GET / HTTP/1.1
Host: localhost:8000
User-Agent: Some custom test header
Accept: */*
Accept-Encoding: gzip, deflate
```

I see no difference in the requests being sent 
I also tested it with the [requests](https://docs.python-requests.org/en/latest/) library which manages to get a 200 response from reddit on all versions.

```
GET / HTTP/1.1
Host: localhost:8000
User-Agent: Some custom test header
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
```
the only change here is that requests also sends a `Connection: keep-alive` header but adding that header to my code above for testing doesn't change the outcome for aiohttp

**Reply by Dreamsorcerer** [reply]
So, the difference may only be in the TLS version or similar (I'm not familiar with how to debug at this low level). Still something to ask the support email in that response, it's unlikely we'll be able to do anything here.

We've had something similar with Tesla's API. They use a firewall (akamai?) and it ended up sometimes blocking requests on really odd details that was impossible to track down (it was so bad, I think they even ended up blocking their own official app at one point).

**Reply by tailoric** [reply]
Yeah I agree, seems like it is something we need to ask reddit directly.
Seems something in their server configurations changed since this problem has been happening for at least 3 weeks (at least that has been when I noticed it)

Though I am not sure if I'll get an answer because I asked on the 19th of September why I was IP blocked (I thought that was the reason at the time) and haven't gotten a response yet.
I will forward another E-Mail and include a link to this discussion and see if I will get an answer.

**Reply by webknjaz** [reply]
@tailoric I think you may be able to see the TLS version and ciphers in the handshake if you record and open it via Wireshark. I don't remember, though, if the beginning of the handshake is encrypted or not, needs testing.
